staging/src/k8s.io/client-go/util/workqueue/priority_queue.go

```diff
@@ -0,0 +1,302 @@
+	t := &PriorityType{
+		minPriority:                defaultMinPriority,
+		maxPriority:                maxPriority,
+		getPriorityFunc:            f,
+		priorityQueue:              map[int][]t{},
+		clock:                      c,
+		dirty:                      set{},
+		processing:                 set{},
+		cond:                       sync.NewCond(&sync.Mutex{}),
+		metrics:                    metrics,
+		unfinishedWorkUpdatePeriod: updatePeriod,
+	}
+
+	// Don't start the goroutine for a type of noMetrics so we don't consume
+	// resources unnecessarily
+	if _, ok := metrics.(noMetrics); !ok {
+		go t.updateUnfinishedWorkLoop()
+	}
+
+	return t
+}
+
+// PriorityType is a work queue (see the package comment).
+type PriorityType struct {
+	minPriority     int
+	maxPriority     int
+	getPriorityFunc GetPriorityFunc
+
+	// queue defines the order in which we will work on items. Every
+	// element of queue should be in the dirty set and not in the
+	// processing set.
+	priorityQueue map[int][]t
+
+	// dirty defines all of the items that need to be processed.
+	dirty set
+
+	// Things that are currently being processed are in the processing set.
+	// These things may be simultaneously in the dirty set. When we finish
+	// processing something and remove it from this set, we'll check if
+	// it's in the dirty set, and if so, add it to the queue.
+	processing set
+
+	cond *sync.Cond
+
+	shuttingDown bool
+	drain        bool
+
+	metrics queueMetrics
+
+	unfinishedWorkUpdatePeriod time.Duration
+	clock                      clock.WithTicker
+}
+
+// Add marks item as needing processing.
+func (q *PriorityType) Add(item interface{}) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	if q.shuttingDown {
+		return
+	}
+	if q.dirty.has(item) {
+		return
+	}
+
+	q.metrics.add(item)
+
+	q.dirty.insert(item)
+	if q.processing.has(item) {
+		return
+	}
+
+	priority := q.getPriorityFunc(item)
+	priority = clipInt(priority, q.minPriority, q.maxPriority)
+	q.priorityQueue[priority] = append(q.priorityQueue[priority], item)
+	q.cond.Signal()
+}
+
+func clipInt(v, min, max int) int {
+	switch {
+	case v < min:
+		return min
+	case v > max:
+		return max
+	default:
+		return v
+	}
+}
+
+// Len returns the current queue length, for informational purposes only. You
+// shouldn't e.g. gate a call to Add() or Get() on Len() being a particular
+// value, that can't be synchronized properly.
+func (q *PriorityType) Len() int {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+
+	return q.lenNoLock()
+}
+func (q *PriorityType) lenNoLock() int {
+	count := 0
+	for i := q.minPriority; i <= q.maxPriority; i++ {
+		count += len(q.priorityQueue[i])
+	}
+	return count
+}
+
+// Get blocks until it can return an item to be processed. If shutdown = true,
+// the caller should end their goroutine. You must call Done with item when you
+// have finished processing it.
+func (q *PriorityType) Get() (item interface{}, shutdown bool) {
```

Comment:
the priority queue should be based on a heap 

https://pkg.go.dev/container/heap#example-package-PriorityQueue

so it can scale well

https://en.wikipedia.org/wiki/Priority_queue

Reply:
Thanks a lot, very good suggestion, I will refactor priority_queue.go based on heap.